### PR TITLE
BINDINGS/GO: Standalone performance benchmark.

### PR DESCRIPTION
**Pull request by petro-rudenko**
## What
GO: Standalone performance benchmark.

### CODE REVIEW

**Comment by petro-rudenko** [comment]
@yosefe ready to review. The solution for CUDA so far is to [register first dummy host buffer](https://github.com/openucx/ucx/pull/7604/commits/35d5174064d8dd194275486a75de26f505ed1b6a). Tried different other solutions, like [Lock OS Thread](https://github.com/golang/go/wiki/LockOSThread) - the same situation, unless we first mapping host buffer. Will check more on this - if found better solution will fix.

File: bindings/go/src/examples/perftest/perftest.go [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+	epParams := &UcpEpParams{}
+	serverAddress, _ := net.ResolveTCPAddr("tcp", fmt.Sprintf("%v:%v", perfTestParams.ip, perfTestParams.port))
+	epParams.SetPeerErrorHandling().SetErrorHandler(func(ep *UcpEp, status UcsStatus) {
+		if status != UCS_ERR_CONNECTION_RESET {
+			fmt.Printf("Endpoint error: %v \n", status.String())
+		}
+	}).SetSocketAddress(serverAddress)
+
+	perfTest.eps[i], err = perfTest.perThreadWorkers[i].NewEndpoint(epParams)
+
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:156]: remove spaceline

File: bindings/go/src/examples/perftest/perftest.go [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+	if err != nil {
+		return err
+	}
+
+	request, err := perfTest.eps[i].FlushNonBlocking(nil)
+
+	if err != nil {
+		return err
+	}
+
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:166]: remove spaceline

File: bindings/go/src/examples/perftest/perftest.go [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+	listenerParams := &UcpListenerParams{}
+	addr, _ := net.ResolveTCPAddr("tcp", fmt.Sprintf("0.0.0.0:%v", perfTestParams.port))
+
+	listenerParams.SetSocketAddress(addr)
+	listenerParams.SetConnectionHandler(func(connRequest *UcpConnectionRequest) {
+		backwardEp, _ := perfTest.perThreadWorkers[0].NewEndpoint((&UcpEpParams{}).
+			SetConnRequest(connRequest).SetPeerErrorHandling())
+
+		// Current connection establishment requires to create backward ep or reject connection.
+		perfTest.backwardEps = append(perfTest.backwardEps, backwardEp)
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:190]: backwardEps -> reverseEps

File: bindings/go/src/examples/perftest/perftest.go [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+}
+
+func initListener() error {
+	var err error
+	listenerParams := &UcpListenerParams{}
+	addr, _ := net.ResolveTCPAddr("tcp", fmt.Sprintf("0.0.0.0:%v", perfTestParams.port))
+
+	listenerParams.SetSocketAddress(addr)
+	listenerParams.SetConnectionHandler(func(connRequest *UcpConnectionRequest) {
+		backwardEp, _ := perfTest.perThreadWorkers[0].NewEndpoint((&UcpEpParams{}).
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:186]: reverseEp

File: bindings/go/src/examples/perftest/perftest.go [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+			SetConnRequest(connRequest).SetPeerErrorHandling())
+
+		// Current connection establishment requires to create backward ep or reject connection.
+		perfTest.backwardEps = append(perfTest.backwardEps, backwardEp)
+		fmt.Println("Got connection starting benchmark")
+	})
+
+	perfTest.listener, err = perfTest.perThreadWorkers[0].NewListener(listenerParams)
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:194]: need to have multiple threads on the receiver
the listener will create reply ep on different workers/threads

File: bindings/go/src/examples/perftest/perftest.go [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+	}
+	fmt.Printf("Started receiver listener on address: %v \n", addr)
+	return nil
+}
+
+func progressWorker(i int) {
+	perfTest.perThreadWorkers[i].Progress()
+}
+
+func submitTagRecvRequest(tid uint) {
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:206]: pls use only active messages in this example

File: bindings/go/src/examples/perftest/perftest.go [hunk 7]
```diff
@@ -0,0 +1,397 @@
...
+	if err := initMemory(); err != nil {
+		return err
+	}
+	perfTest.perThreadWorkers = make([]*UcpWorker, 1)
+	initWorker(0)
+	if err := initListener(); err != nil {
+		return err
+	}
+
+	if perfTestParams.protocol == "am" {
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:253]: only am should be supported, remove "protocol"

File: bindings/go/src/examples/perftest/perftest.go [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+		for t := uint(0); t < perfTestParams.numThreads; t += 1 {
+			tid := t
+			perfTest.perThreadWorkers[0].SetAmRecvHandler(t, UCP_AM_FLAG_WHOLE_MSG,
+				func(header unsafe.Pointer, headerSize uint64, data *UcpAmData, replyEp *UcpEp) UcsStatus {
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:257]: maybe define it separately?

File: bindings/go/src/examples/perftest/perftest.go [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+
+		for perfTest.numCompletedRequests < perfTestParams.numThreads {
+			progressWorker(0)
+		}
+	}
+	close()
+	return nil
+}
+
+func iterPerThread(i uint, t uint) {
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:291]: can we add comment, or specific name, to distinguish client and server functions?
for example: clientThreadDoIter

File: bindings/go/src/examples/perftest/perftest.go [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+	perfTest.completionTime[t] = time.Since(start)
+	request.Close()
+
+	if i%perfTestParams.printIter == 0 {
+		printPerThreadStatistics(i, t)
+	}
+	perfTest.wg.Done()
+}
+
+func startClient() error {
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:314]: clientStart
serverStart
...

File: buildlib/pr/go/go-test.yml [hunk 11]
```diff
@@ -52,3 +52,17 @@ jobs:
...
+        az_init_modules
+        try_load_cuda_env
+        az_module_load dev/go-latest
+        go_port=$((20000 + $RANDOM % 10000))
+        args="-p=$go_port"
+        if [ "${{ parameters.name }}" == "gpu" ]; then
+          args="-m=cuda"
+        fi
+        make -C build/bindings/go run-perftest ARGS="$args" &
+        make -C build/bindings/go run-perftest ARGS="$args -i=localhost"
```
Comment (yosefe) [inline buildlib/pr/go/go-test.yml:67]: maybe run this step w/o makefile?

File: bindings/go/src/examples/perftest/perftest.go [hunk 12]
```diff
@@ -239,65 +225,66 @@ func close() {
 	}
 }
 
-func startServer() error {
+func serverAmRecvHandler(header unsafe.Pointer, headerSize uint64, data *UcpAmData, replyEp *UcpEp) UcsStatus {
+	tid := *(*uint)(header)
+	if data.IsDataValid() {
+		atomic.AddUint32(&perfTest.numCompletedRequests, 1)
+		return UCS_OK
+	} else {
```
Comment (yosefe) [inline bindings/go/src/examples/perftest/perftest.go:233]: minor: maybe remove "else" since have "return"?

File: bindings/go/Makefile.am [hunk 13]
```diff
@@ -37,6 +37,14 @@ bench: $(GOTMPDIR)
 	$(GO) env -w GO111MODULE=off ; \
 	LD_LIBRARY_PATH=$(UCX_SOPATH):${LD_LIBRARY_PATH} $(GO) test -v -bench=.
 
-all: build
+build-perftest: $(GOTMPDIR)
```
Comment (yosefe) [inline bindings/go/Makefile.am:40]: 1. the Makefile target name should be the resulting binary name, for example "goperftest"
2. all targets that do not create output (e.g run-perftest, all) should be marked as phony:
`.PHONY: all run_perftest test bench`
3. i think the target "build" should also be changed to the name of the resulting binary

Comment (petro-rudenko) [reply bindings/go/Makefile.am:40]: `build` just builds go bindings but does not result in some lib or binary.

Comment (yosefe) [reply bindings/go/Makefile.am:40]: doesn't it create any output files in GOOBJDIR?

Comment (petro-rudenko) [reply bindings/go/Makefile.am:40]: Just temporary files.

Comment (yosefe) [reply bindings/go/Makefile.am:40]: ok then

File: bindings/go/Makefile.am [hunk 14]
```diff
@@ -37,14 +37,19 @@ bench: $(GOTMPDIR)
...
 	cd $(abs_top_srcdir)/bindings/go/src/examples/perftest ;\
 	$(GO) build -o ${GOTMPDIR}/goperftest
 
 run-perftest:
 	cd $(abs_top_srcdir)/bindings/go/src/examples/perftest ;\
 	LD_LIBRARY_PATH=$(UCX_SOPATH):${LD_LIBRARY_PATH} ${GOTMPDIR}/goperftest ${ARGS}
 
-all: build-perftest build
+install-exec-hook:
+	cp ${GOTMPDIR}/goperftest $(DESTDIR)$(bindir)
```
Comment (yosefe) [inline bindings/go/Makefile.am:49]: minor: it's more customary to use `$(INSTALL)` command (though is basically same as cp)

File: buildlib/pr/go/go-test.yml [hunk 15]
```diff
@@ -39,7 +39,7 @@ jobs:
         ./autogen.sh
         mkdir build
         cd build
-        ../contrib/configure-devel --prefix=$(Build.Repository.LocalPath)/install \
+        ../contrib/configure-devel --prefix=$(Build.ArtifactStagingDirectory)/ucx-install \
```
Comment (yosefe) [inline buildlib/pr/go/go-test.yml:42]: why using ArtifactStagingDirectory? is it for OOS build?

Comment (petro-rudenko) [reply buildlib/pr/go/go-test.yml:42]: Yes for OOS test

Comment (yosefe) [reply buildlib/pr/go/go-test.yml:42]: maybe comment it; when looking at this it seems we're trying create artifacts but it's not really the purpose

File: buildlib/pr/go/go-test.yml [hunk 16]
```diff
@@ -63,6 +63,6 @@ jobs:
         if [ "${{ parameters.name }}" == "gpu" ]; then
           args="-m=cuda"
         fi
-        LD_LIBRARY_PATH=$(Build.ArtifactStagingDirectory)/ucx-install/lib/:$LD_LIBRARY_PATH $(Build.ArtifactStagingDirectory)/ucx-install/bin/goperftest $args &
-        LD_LIBRARY_PATH=$(Build.ArtifactStagingDirectory)/ucx-install/lib/:$LD_LIBRARY_PATH $(Build.ArtifactStagingDirectory)/ucx-install/bin/goperftest $args -i=localhost
+        LD_LIBRARY_PATH=$(Build.ArtifactStagingDirectory)/ucx-install/lib/:$LD_LIBRARY_PATH $(Agent.TempDirectory)/ucx-$(Build.BuildId)/bin/goperftest $args &
```
Comment (yosefe) [inline buildlib/pr/go/go-test.yml:66]: nice.

**Review by yosefe — APPROVED** [review]

**Pull request merged**